File: 02 clase cadenas/py/cadena.py
class Cadena:
    def __init__(self, cadena):
        self.cadena = cadena

    def get_cadena(self):  # Obtiene el valor asignado a la cadena
        return self.cadena

    # Manipulación de cadenas
    def concatenar(self, otra_cadena):  # Unir una cadena con otra
        self.cadena += otra_cadena

    def invertir(self):  # Invertir los caracteres de la cadena
        self.cadena = self.cadena[::-1]

    # Comparaciones
    def es_vacia(self):  # Compara si la cadena no tiene caracteres
        return len(self.cadena) == 0

    def contiene(self, subcadena):  # Compara si la subCadena existe dentro de la cadena
        return subcadena in self.cadena

    def comienza_con(self, prefijo):  # Compara si la cadena comienza con el prefijo "caracter"
        return self.cadena.startswith(prefijo)

    def termina_con(self, sufijo):  # Compara si la cadena termina con el sufijo "caracter"
        return self.cadena.endswith(sufijo)

    # Otras operaciones
    def longitud(self):  # Obtiene la cantidad de caracteres que tiene la cadena
        return len(self.cadena)

    def a_mayusculas(self):  # Obtiene la cadena transformada en MAYÚSCULA
        return self.cadena.upper()

    def a_minusculas(self):  # Obtiene la cadena transformada en minúscula
        return self.cadena.lower()

# Ejemplo de uso
cadena = Cadena('')

def ejecutar(metodo, valor_input=''):
    resultado = ''
    if metodo == 'getCadena':
        resultado = f'Cadena: {cadena.get_cadena()}'
    elif metodo == 'concatenar':
        cadena.concatenar(valor_input)
        resultado = f'Cadena actual: {cadena.get_cadena()}'
    elif metodo == 'invertir':
        cadena.invertir()
        resultado = f'Cadena invertida: {cadena.get_cadena()}'
    elif metodo == 'esVacia':
        resultado = f'¿Es vacía?: {cadena.es_vacia()}'
    elif metodo == 'contiene':
        resultado = f'¿Contiene "{valor_input}"?: {cadena.contiene(valor_input)}'
    elif metodo == 'comienzaCon':
        resultado = f'¿Comienza con "{valor_input}"?: {cadena.comienza_con(valor_input)}'
    elif metodo == 'terminaCon':
        resultado = f'¿Termina con "{valor_input}"?: {cadena.termina_con(valor_input)}'
    elif metodo == 'longitud':
        resultado = f'Longitud: {cadena.longitud()}'
    elif metodo == 'aMayusculas':
        resultado = f'A mayúsculas: {cadena.a_mayusculas()}'
    elif metodo == 'aMinusculas':
        resultado = f'A minúsculas: {cadena.a_minusculas()}'
    else:
        resultado = 'Método no reconocido.'

    return resultado

# Ejemplo de uso
print(ejecutar('concatenar', 'Hola '))
print(ejecutar('getCadena'))
print(ejecutar('invertir'))
print(ejecutar('longitud'))

File: 02 clase cadenas/js/cadena.js
class Cadena {
  constructor(cadena) {
    this.cadena = cadena;
  }

  getCadena() {
    // Obtiene el valor asignado a la cadena
    return this.cadena;
  }

  // Manipulación de cadenas
  concatenar(otraCadena) {
    // Une una cadena con otra
    this.cadena += otraCadena;
  }

  invertir() {
    // Invierte los caracteres de la cadena
    let cadenaInvertida = '';
    for (let i = this.cadena.length - 1; i >= 0; i--) {
      cadenaInvertida += this.cadena[i];
    }
    this.cadena = cadenaInvertida;
  }

  // Comparaciones
  esVacia() {
    // Compara si la cadena no tiene caracteres
    return this.cadena.length === 0;
  }

  contiene(subcadena) {
    // Compara si la subcadena existe dentro de la cadena
    return this.cadena.includes(subcadena);
  }

  comienzaCon(prefijo) {
    // Compara si la cadena comienza con el prefijo
    return this.cadena.startsWith(prefijo);
  }

  terminaCon(sufijo) {
    // Compara si la cadena termina con el sufijo
    return this.cadena.endsWith(sufijo);
  }

  // Otras operaciones
  longitud() {
    // Obtiene la cantidad de caracteres que tiene la cadena
    return this.cadena.length;
  }

  aMayusculas() {
    // Obtiene la cadena transformada en MAYÚSCULA
    return this.cadena.toUpperCase();
  }

  aMinusculas() {
    // Obtiene la cadena transformada en minúscula
    return this.cadena.toLowerCase();
  }

  // Transformación de números a números romanos
  static numeroARomano(numero) {
    const romanos = [
      ['M', 1000], ['CM', 900], ['D', 500], ['CD', 400],
      ['C', 100], ['XC', 90], ['L', 50], ['XL', 40],
      ['X', 10], ['IX', 9], ['V', 5], ['IV', 4], ['I', 1]
    ];
    let resultado = '';
    for (const [letra, valor] of romanos) {
      while (numero >= valor) {
        resultado += letra;
        numero -= valor;
      }
    }
    return resultado;
  }

  // Generación de Texto Aleatorio
  static generarTextoAleatorio(cantidad) {
    const caracteres = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 ';
    let resultado = '';
    for (let i = 0; i < cantidad; i++) {
      resultado += caracteres.charAt(Math.floor(Math.random() * caracteres.length));
    }
    return resultado;
  }

  // Transformación de ASCII a Texto
  static asciiATexto(ascii) {
    return String.fromCharCode(ascii);
  }
}

let cadena = new Cadena('');

// Función para manejar la ejecución desde la interfaz
function ejecutar(metodo) {
  const input = document.getElementById('inputCadena');
  const resultadoDiv = document.getElementById('resultado');
  const valorInput = input.value;

  switch (metodo) {
    case 'getCadena':
      resultadoDiv.innerHTML = `Cadena: ${cadena.getCadena()}`;
      break;
    case 'concatenar':
      cadena.concatenar(valorInput);
      resultadoDiv.innerHTML = `Cadena actual: ${cadena.getCadena()}`;
      break;
    case 'invertir':
      cadena.invertir();
      resultadoDiv.innerHTML = `Cadena invertida: ${cadena.getCadena()}`;
      break;
    case 'esVacia':
      resultadoDiv.innerHTML = `¿Es vacía?: ${cadena.esVacia()}`;
      break;
    case 'contiene':
      resultadoDiv.innerHTML = `¿Contiene '${valorInput}'?: ${cadena.contiene(valorInput)}`;
      break;
    case 'comienzaCon':
      resultadoDiv.innerHTML = `¿Comienza con '${valorInput}'?: ${cadena.comienzaCon(valorInput)}`;
      break;
    case 'terminaCon':
      resultadoDiv.innerHTML = `¿Termina con '${valorInput}'?: ${cadena.terminaCon(valorInput)}`;
      break;
    case 'longitud':
      resultadoDiv.innerHTML = `Longitud: ${cadena.longitud()}`;
      break;
    case 'aMayusculas':
      resultadoDiv.innerHTML = `A mayúsculas: ${cadena.aMayusculas()}`;
      break;
    case 'aMinusculas':
      resultadoDiv.innerHTML = `A minúsculas: ${cadena.aMinusculas()}`;
      break;
    case 'numeroARomano':
      resultadoDiv.innerHTML = `Número a romano: ${Cadena.numeroARomano(parseInt(valorInput))}`;
      break;
    case 'generarTextoAleatorio':
      resultadoDiv.innerHTML = `Texto aleatorio: ${Cadena.generarTextoAleatorio(parseInt(valorInput))}`;
      break;
    case 'asciiATexto':
      resultadoDiv.innerHTML = `ASCII a texto: ${Cadena.asciiATexto(parseInt(valorInput))}`;
      break;
    default:
      resultadoDiv.innerHTML = 'Método no reconocido.';
  }
}